File: THCS/2022_2023/YenBai_2223/Cau4/SPSEQ.py
def longest_cone_subsequence(arr):
    n = len(arr)
    
    if n < 3:
        return 0  # Không đủ số phần tử để tạo thành dãy hình nón

    # Tạo mảng tăng dần và giảm dần
    inc = [0] * n
    dec = [0] * n

    # Tính độ dài chuỗi con tăng dần
    for i in range(n):
        for j in range(i):
            if arr[i] > arr[j] and arr[i] != arr[j]:
                inc[i] = max(inc[i], inc[j] + 1)

    # Tính độ dài chuỗi con giảm dần
    for i in range(n - 1, -1, -1):
        for j in range(i + 1, n):
            if arr[i] > arr[j] and arr[i] != arr[j]:
                dec[i] = max(dec[i], dec[j] + 1)

    max_length = 0

    # Tính độ dài chuỗi con hình nón
    for i in range(n):
        # x + 1 phần tử tăng, x + 1 phần tử giảm => m = 2x + 1
        if inc[i] + dec[i] >= 1:  # Điều kiện phải có ít nhất 1 phần tử tăng và giảm
            current_length = inc[i] + dec[i] + 1  # +1 cho phần tử giữa
            if current_length % 2 == 1:  # Đảm bảo độ dài là số lẻ
                max_length = max(max_length, current_length)

    return max_length

def main():
    with open("SPSEQ.INP", "r") as infile:
        n = int(infile.readline().strip())
        arr = list(map(int, infile.readline().strip().split()))

    result = longest_cone_subsequence(arr)

    with open("SPSEQ.OUT", "w") as outfile:
        outfile.write(str(result) + '\n')

if __name__ == "__main__":
    main()

File: THCS/2022_2023/YenBai_2223/Cau4/SPSEQ.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

// Hàm tìm độ dài dãy con hình nón dài nhất
int longestConeSubsequence(const vector<int>& arr) {
    int n = arr.size();
    
    if (n < 3) return 0; // Không đủ số phần tử để tạo thành dãy hình nón

    vector<int> inc(n, 1);
    vector<int> dec(n, 1);

    // Tính độ dài chuỗi con tăng dần
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < i; ++j) {
            if (arr[i] > arr[j] && arr[i] != arr[j]) {
                inc[i] = max(inc[i], inc[j] + 1);
            }
        }
    }

    // Tính độ dài chuỗi con giảm dần
    for (int i = n - 1; i >= 0; --i) {
        for (int j = n - 1; j > i; --j) {
            if (arr[i] > arr[j] && arr[i] != arr[j]) {
                dec[i] = max(dec[i], dec[j] + 1);
            }
        }
    }

    int max_length = 0;

    // Tính độ dài chuỗi con hình nón
    for (int i = 0; i < n; ++i) {
        if (inc[i] + dec[i] >= 1) { // Điều kiện phải có ít nhất 1 phần tử tăng và giảm
            int current_length = inc[i] + dec[i] - 1; // -1 vì phần tử giữa được tính 2 lần
            if (current_length % 2 == 0) { // Đảm bảo độ dài là số lẻ
                current_length++; // Nếu current_length là số chẵn, tăng lên 1 để thành số lẻ
            }
            max_length = max(max_length, current_length);
        }
    }

    return max_length;
}

int main() {
    ifstream infile("SPSEQ.INP");
    ofstream outfile("SPSEQ.OUT");
    
    int n;
    infile >> n;

    vector<int> arr(n);
    for (int i = 0; i < n; ++i) {
        infile >> arr[i];
    }

    int result = longestConeSubsequence(arr);
    outfile << result << endl;

    infile.close();
    outfile.close();

    return 0;
}
